---
title: "Ultimate_SnRNA"
author: "Hugues Escoffier"
output: html_document
---

# SnRNA-seq 

The sequencing data preprocessing steps, including base calling, mapping and read counting, have been done using Cell Ranger. 
Based on Zhisong He & Barbara Treutlein Analysis (https://github.com/quadbiolab)

## 0. Set Up 

### 0.1 Libraries 

```{r}
# CRAN 
library(Seurat)
library(SeuratDisk)
library(circlize)
library(harmony)
library(dplyr)
library(tidyr)
library(tidyverse)
library(patchwork)
library(stats)
library(ggplot2)
library(ggrepel)
library(tibble)
library(monocle3)
library(SeuratWrappers)
library(clustree)
library(anndata)
library(reticulate)
library(renv)
# BioConductor 
library(destiny)
library(ComplexHeatmap)
library(biomaRt)
library(DESeq2)
library(scran)
# Devtools 
library(presto)
library(nichenetr)
library(loomR)
library(chorddiag)
```

### 0.2 Palettes 

```{r}
vangogh <- c("#003366", "#003399", "#0033CC", "#0066CC", "#0099CC", "#00CCCC", "#33CCCC", "#66CCCC", "#66CC99", "#99CC99", "#CCCC99", "#CCCC66", "#FFCC00", "#FF9900", "#FF6600", "#FF3300", "#CC0033", "#CC0066", "#CC0099", "#CC33FF", "#9933FF", "#6600FF")
matisse <- c("#D6BCC0", "#FFCD00", "#009BDE", "#FF5733", "#8B1A1A", "#81C784", "#FA8072", "#009688", "#F44336", "#FFA07A", "#3F51B5", "#F9A8D4", "#006400", "#BA55D3", "#6600FF")
distinct <- c('#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#fffac8', '#800000', '#aaffc3', '#808000', '#ffd8b1', '#000075', '#808080')
stepped <- c("#990f26", "#b33e52", "#cc7a88", "#e6b8bf", "#99600f", "#b3823e", "#ccaa7a", "#e6d2b8", "#54990f", "#78b33e", "#a3cc7a", "#cfe6b8", "#0f8299", "#3e9fb3", "#7abecc", "#b8dee6", "#3d0f99", "#653eb3", "#967acc", "#c7b8e6", "#333333", "#666666")
impress <- c("#f0c8a0","#e0b0ff","#ffbf00","#8b4513","#87ceeb","#ffc0cb","#b0e0e6","#f0e68c","#6b8e23",
                      # "#dda0dd",
                      "#ff7f50","#fafad2","#d2b48c","#ff6347",
                      # "#f5f5dc",
                      "#e6e6fa","#bdb76b","#b22222","#cd853f","#00ff7f")
laputa <- c('#403369FF', "#AE93BEFF", "#B4DAE5FF", "#F0D77BFF")
```

## 1. Create Seurat Object 

```{r}
counts <- Read10X(data.dir = "data/WTSO/filtered_feature_bc_matrix")
# seurat.rna <- CreateSeuratObject(counts$`Gene Expression`, project="D2")
seurat.rna <- CreateSeuratObject(counts, project="Ctl-Pltr")
```

## 2. Quality Control 
We filter : 
* Cells with too few genes detected. _(representing cells which are not sequenced deep enough)_
* Cells with too many genes detected. _(corresponding to multiplets)_
* Cells with high mitochondrial transcript percentage.

```{r}
# Add mitochondrial transcript percentage
seurat.rna[["percent.mt"]] <- PercentageFeatureSet(seurat.rna, pattern = "^mt-")
# Visualize 
VlnPlot(seurat.rna, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size=0)
FeatureScatter(seurat.rna, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(seurat.rna, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
# The number of detected genes and number of detected transcripts are well correlated across cells while mitochondrial transcript percentage is not.
```

```{r}
# Same thresholds as Lin et al. 
seurat.rna <- subset(seurat.rna, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 0.8)
```

## 3. Normalization 
Since the amount of captured RNA can vary between cells, it is not appropriate to compare the number of captured transcripts for individual genes directly. To ensure that gene expression levels can be accurately compared across different cells, it is necessary to perform a normalization step. This step is aimed at making gene expression levels between cells comparable.

```{r}
seurat.rna <- NormalizeData(seurat.rna)
```

## 4. Feature selection
Identification of highly variable features/genes, which are genes with the most varied expression levels across cells.

```{r}
# There are no defined values in the literature. Generally we choose values between 2000 and 5000. The results do not differ significantly between these values. 
seurat.rna <- FindVariableFeatures(seurat.rna, nfeatures = 3000)
```

```{r}
# Visualize
top_features <- head(VariableFeatures(seurat.rna), 20)
plot1 <- VariableFeaturePlot(seurat.rna)
plot1
LabelPoints(plot = plot1, points = top_features, repel = TRUE)
```

## 5. Scaling 
If no data transformation is performed, the analysis could be biased towards highly expressed genes since different genes have varying base expression levels and distributions. This is an undesirable outcome as we want our analysis to consider all genes equally. To address this issue, we apply scaling to the data using selected features, which is a common practice in various data science fields. This ensures that the contribution of each gene to the analysis is appropriately accounted for, regardless of their expression levels.

```{r}
seurat.rna <- ScaleData(seurat.rna, vars.to.regress = c("nFeature_RNA", "percent.mt")) # Remove unwanted sources of variation
```

## 3-5. SCTransform 
Steps 3 to 5 can be replaced by a `SCTransform` step

```{r}
seurat.rna <- SCTransform(seurat.rna,
                      vars.to.regress = c("nFeature_RNA", "percent.mt"),
                      variable.features.n = 3000)
```


## 6. PCA

```{r}
seurat.rna <- RunPCA(seurat.rna, npcs = 50)
ElbowPlot(seurat.rna, ndims = ncol(Embeddings(seurat.rna, "pca")))
```

One would assume that the first phase of the curve represents the 'real' signal related to biological differences between cell populations, while the second phase mostly represents technical variation or the stochastic nature of individual cells.
* It is very difficult to precisely define the elbow point or turning point of the curve, as it is usually not a perfect elbow.
* Higher-ranked PCs do explain more variation than lower-ranked PCs, but more explained variations does not necessarily mean higher information content. Sometimes there is interesting but weak signal buried in the noise and therefore as part of lower-ranked PCs.

## 7. Non-linear dimension reduction 
TSNE provides great visualization when cells form distinct cell groups, while UMAP perserves trajectory-like structure better when data contains 'continuum'.

```{r}
seurat.rna <- RunTSNE(seurat.rna, dims = 1:15)
seurat.rna <- RunUMAP(seurat.rna, dims = 1:15)

TSNEPlot(seurat.rna)
UMAPPlot(seurat.rna)
```

```{r}
FeaturePlot(seurat.rna, c("Myh1","Myh2","Myh4","Myh7","Xist","Uty","Col22a1","Pam","Ttn"),
ncol=3, reduction = "tsne")
FeaturePlot(seurat.rna, c("Myh1","Myh2","Myh4","Myh7","Xist","Uty","Col22a1","Pam","Ttn"),
ncol=3, reduction = "umap")
```

## Save 

```{r}
saveRDS(seurat.rna, file="save/seurat_ctl_pl")
```

## 8. Integration 
Benchmarking of integration tools : 
* A benchmark of batch-effect correction methods for single-cell RNA sequencing data (https://doi.org/10.1186/s13059-019-1850-9)
* Benchmarking atlas-level data integration in single-cell genomics (https://doi.org/10.1038/s41592-021-01336-8)

### 8.0 Load 

```{r}
seurat_ctl <- readRDS("save/seurat_ctl")
seurat_denerv <- readRDS("save/seurat_denerv")
seurat_d2 <- readRDS("save/seurat_d2")
seurat_teno <- readRDS("save/seurat_teno")
seurat_ctl_pl <- readRDS("save/seurat_ctl_pl")
```

### 8.1 Merge

```{r}
seurat.merge <- merge(seurat_ctl, c(seurat_denerv, seurat_d2, seurat_ctl_pl)) %>%
# seurat.merge <- merge(seurat_ctl_pl, c(seurat_teno)) %>%
  FindVariableFeatures(nfeatures = 3000) %>%
  ScaleData() %>%
  RunPCA(npcs = 50) %>%
  RunUMAP(dims = 1:20)

# Rename 
seurat.merge$orig.ident[seurat.merge$orig.ident == "Ctl"] <- "Ctl-TA"
seurat.merge$orig.ident[seurat.merge$orig.ident == "Ctl-Pl"] <- "Ctl-Pltr"
seurat.merge$orig.ident[seurat.merge$orig.ident == "D2"] <- "D-Pltr"
seurat.merge$orig.ident[seurat.merge$orig.ident == "Denerv"] <- "D-TA"

DimPlot(seurat.merge, group.by="orig.ident", cols = matisse) & NoAxes()
ggsave("PLOT/4D_Merge.tiff", device = grDevices::tiff)
FeaturePlot(seurat.merge, c("Myh1","Myh2","Myh4","Musk"), ncol=2, pt.size = 0.1) & NoAxes()
ggsave("PLOT/4D_Merge_Expression.tiff", device = grDevices::tiff)
```

### 8.2-A Using Seurat 
_(https://doi.org/10.1016/j.cell.2019.05.031)_

```{r}
seurat_objs <- list(seurat_ctl, seurat_denerv, seurat_d2, seurat_ctl_pl)
# seurat_objs <- list(seurat_ctl_pl, seurat_teno)
anchors <- FindIntegrationAnchors(object.list = seurat_objs, dims = 1:30)
seurat.integrate <- IntegrateData(anchors, dims = 1:30)
```

```{r}
seurat.integrate <- ScaleData(seurat.integrate)
seurat.integrate <- RunPCA(seurat.integrate, npcs = 50)
ElbowPlot(seurat.integrate, ndims = ncol(Embeddings(seurat.integrate, "pca")))
```

```{r}
seurat.integrate <- RunUMAP(seurat.integrate, dims = 1:20)
```

```{r}
DefaultAssay(seurat.integrate) <- "RNA"
UMAPPlot(seurat.integrate, group.by="orig.ident")
FeaturePlot(seurat.integrate, c("Myh1","Myh2","Myh4","Col22a1"), ncol=2, pt.size = 0.1)

```

### 8.2-B Using Harmony 
_(https://doi.org/10.1038/s41592-019-0619-0)_

```{r}
seurat.harm <- merge(seurat_ctl, c(seurat_denerv, seurat_d2, seurat_ctl_pl)) %>%
# seurat.harm <- merge(seurat_ctl_pl, c(seurat_teno)) %>%
  FindVariableFeatures(nfeatures = 3000) %>%
  ScaleData() %>%
  RunPCA(npcs = 50)

seurat.harm$orig.ident[seurat.harm$orig.ident == "Ctl"] <- "Ctl-TA"
seurat.harm$orig.ident[seurat.harm$orig.ident == "Ctl-Pl"] <- "Ctl-Pltr"
seurat.harm$orig.ident[seurat.harm$orig.ident == "D2"] <- "D-Pltr"
seurat.harm$orig.ident[seurat.harm$orig.ident == "Denerv"] <- "D-TA"

seurat.harm <- RunHarmony(seurat.harm, group.by.vars = "orig.ident", dims.use = 1:25, max.iter.harmony = 50)
seurat.harm <- RunUMAP(seurat.harm, reduction = "harmony", dims = 1:25)
```

```{r}
DefaultAssay(seurat.harm) <- "RNA"
UMAPPlot(seurat.harm, group.by="orig.ident", cols = matisse) & NoAxes()
ggsave("PLOT/4D_Int.tiff", device = grDevices::tiff)
UMAPPlot(seurat.harm, group.by="orig.ident", split.by = "orig.ident", cols=matisse, ncol = 2) & NoAxes()
ggsave("PLOT/4D_Int_split.tiff", device = grDevices::tiff)
FeaturePlot(seurat.harm, c("Myh1","Myh2","Myh4","Musk"), ncol=2, pt.size = 0.1) & NoAxes()
ggsave("PLOT/4D_Int_Expression.tiff", device = grDevices::tiff)
```

```{r}
saveRDS(seurat.integrate, file="save/seurat_4integrate")
```

## 9. Clustering 

```{r}
# seurat.rna <- seurat.integrate
seurat.rna <- seurat.harm

# Rename 
seurat.rna$orig.ident[seurat.rna$orig.ident == "Ctl"] <- "Ctl-TA"
seurat.rna$orig.ident[seurat.rna$orig.ident == "Ctl-Pl"] <- "Ctl-Pltr"
seurat.rna$orig.ident[seurat.rna$orig.ident == "D2"] <- "D-Pltr"
seurat.rna$orig.ident[seurat.rna$orig.ident == "Denerv"] <- "D-TA"

# Add type
seurat.rna$status <- NA 
seurat.rna$status[seurat.rna$orig.ident == "Ctl-TA"] <- "C"
seurat.rna$status[seurat.rna$orig.ident == "Ctl-Pltr"] <- "C"
seurat.rna$status[seurat.rna$orig.ident == "D-Pltr"] <- "D"
seurat.rna$status[seurat.rna$orig.ident == "D-TA"] <- "D"

# Add muscle
seurat.rna$muscle <- NA 
seurat.rna$muscle[seurat.rna$orig.ident == "Ctl-TA"] <- "TA"
seurat.rna$muscle[seurat.rna$orig.ident == "Ctl-Pltr"] <- "Pltr"
seurat.rna$muscle[seurat.rna$orig.ident == "D-Pltr"] <- "Pltr"
seurat.rna$muscle[seurat.rna$orig.ident == "D-TA"] <- "TA"
```

```{r}
DefaultAssay(seurat.rna) <- "RNA"
# seurat.rna <- FindNeighbors(seurat.rna, dims = 1:ncol(Embeddings(seurat.rna)))
seurat.rna <- FindNeighbors(seurat.rna, dims = 1:ncol(Embeddings(seurat.rna)), reduction = "harmony")
seurat.rna <- FindClusters(seurat.rna, resolution = 0.6)

DimPlot(seurat.rna, reduction = "umap", label = TRUE) & NoAxes()
DimPlot(seurat.rna, reduction = "umap", group.by = "orig.ident", label = FALSE, cols=matisse) & NoAxes()
UMAPPlot(seurat.rna, group.by="orig.ident", split.by = "orig.ident", cols=matisse, ncol=2)  & NoAxes()
DimPlot(seurat.rna, reduction = "umap", group.by="status", label = FALSE, cols = c("#1E488F", "#FF6F61")) & NoAxes()
ggsave("PLOT/4D_Int_Status.tiff", device = grDevices::tiff)
DimPlot(seurat.rna, reduction = "umap", group.by="muscle", label = FALSE, cols = c("#1E488F", "#FF6F61")) & NoAxes() 
ggsave("PLOT/4D_Int_Muscle.tiff", device = grDevices::tiff)

```

```{r}
# Clustree 
DefaultAssay(seurat.rna) <- "RNA"
resolution.range <- seq(from = 0, to = 1, by = 0.2)
seurat.rna.clustree <- FindClusters(object = seurat.rna, resolution = resolution.range)
head(seurat.rna.clustree@meta.data)
clustree(seurat.rna.clustree, prefix = 'RNA_snn_res.0.')
```

## 10. Annotate cells 

```{r}
DefaultAssay(seurat.rna) <- "RNA"
input_data.markers <- FindAllMarkers(seurat.rna, min.pct = 0.25, logfc.threshold = 0.25, only.pos = TRUE)
input_data.markers %>% group_by(cluster) %>% slice_max(n = 5, order_by = avg_log2FC)
input_data.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC) -> topX
# DoHeatmap(seurat.rna, features = topX$gene, group.colors = stepped, hjust = 0, raster = FALSE, lines.width = 5, size = 3) + NoLegend()
```

```{r}
new_ident <- setNames(c("IIb",
                        "Ribo",
                        "IIx_1",
                        "D_IIb",
                        "FAPs",
                        "IIa",
                        "IIx_2",
                        "D_IIx",
                        "Macrophage",
                        "SmoothM",
                        "Tenocytes",
                        "MTJ",
                        "Endothelial",
                        "MuSC",
                        "NMJ",
                        "O_1",
                        "O_2",
                        "Adipocytes"),
                      levels(seurat.rna))
seurat.rna <- RenameIdents(seurat.rna, new_ident)

DimPlot(seurat.rna, reduction = "umap", label = TRUE, cols = stepped
        # cols = ghibli_palettes$LaputaMedium
        ) + NoLegend() + NoAxes()

seurat.rna$celltype <- setNames(rep(c("IIb",
                                      "Ribo",
                                      "IIx_1",
                                      "D_IIb",
                                      "FAPs",
                                      "IIa",
                                      "IIx_2",
                                      "D_IIx",
                                      "Macrophage",
                                      "SmoothM",
                                      "Tenocytes",
                                      "MTJ",
                                      "Endothelial",
                                      "MuSC",
                                      "NMJ",
                                      "O_1",
                                      "O_2",
                                      "Adipocytes"), c(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1)), 
                                c(0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17))[as.character(seurat.rna$RNA_snn_res.0.6)]
```

```{r}
seurat.rna <- subset(seurat.rna, subset = RNA_snn_res.0.6 %in% c(0,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17))
```

```{r}
seurat.rna$gtype <- factor(setNames(c("IIb",
                                      "IIx",
                                      "D_IIb",
                                      "FAPs",
                                      "IIa",
                                      "IIx",
                                      "D_IIx",
                                      "Macrophages",
                                      "SmoothM",
                                      "Tenocytes",
                                      "MTJ",
                                      "Endothelial",
                                      "MuSC",
                                      "NMJ",
                                      "Other",
                                      "Other",
                                      "Adipocytes"),
                                 levels(seurat.rna@active.ident))[seurat.rna@active.ident],
                        levels=c("IIx","IIa", "IIb", "D_IIb", "D_IIx", "MTJ", "NMJ", "FAPs", "Macrophages","SmoothM", "Tenocytes", "Endothelial", "MuSC", "Adipocytes", "Other"))
DimPlot(seurat.rna, reduction = "umap", group.by = "gtype", cols = stepped, label = T) & NoAxes()
ggsave("PLOT/4D_Int_gtype.tiff", device = grDevices::tiff)
```

```{r}
FeaturePlot(seurat.rna, c("Ttn", "Myh1", "Myh2","Myh9"), ncol = 2)
FeaturePlot(seurat.rna, c("Cntfr", "Agbl1", "Kcnn2","Lmcd1"), ncol = 2)
FeaturePlot(seurat.rna, c("Myh2"))
FeaturePlot(seurat.rna, c("Enah", "Hectd2os", "Actn2","Ldlrad3"), ncol = 2)
VlnPlot(seurat.rna, features = c("Stxbp4", "Aldh1a1", "Cep85l","Ppara"), pt.size = 0, ncol = 2)
FeaturePlot(seurat.rna, features = c("Myh1", "Myh2"), blend = TRUE, ncol = 2)
FeatureScatter(seurat.rna, feature1 = "Myh1", feature2 = "Myh2", slot = "counts", group.by = "muscle", cols = matisse, pt.size = 0.2)
ggsave("PLOT/4D_Int_Scatter_12.tiff", device = grDevices::tiff)
FeatureScatter(seurat.rna, feature1 = "Myh4", feature2 = "Myh1", slot = "counts", group.by = "muscle", cols = matisse, pt.size = 0.2)
ggsave("PLOT/4D_Int_Scatter_41.tiff", device = grDevices::tiff)
FeatureScatter(seurat.rna, feature1 = "Myh4", feature2 = "Myh2", slot = "counts", group.by = "muscle", cols = matisse, pt.size = 0.2)
ggsave("PLOT/4D_Int_Scatter_42.tiff", device = grDevices::tiff)
```

## 11. Composition
Check the difference between two conditions is the composition of cell types or cell states. 

### 11.1 Visualize the compostions 

```{r}
freq <- table(seurat.rna$gtype, seurat.rna$status)
prop <- apply(freq,2,function(x) x/sum(x))

layout(matrix(1:3,nrow=1)); par(mar=c(8,5,1,1))
barplot(freq, col=colors <- stepped,
        border=NA, las=2, ylab="Frequency", cex.names = 0.8)
barplot(prop, col=colors <- stepped,
        border=NA, las=2, ylab="Proportion", cex.names = 0.8)
plot.new()
legend("left", fill=stepped, legend=rownames(freq), bty="n")
ggsave("PLOT/4D_Comp_muscle.tiff", device = grDevices::tiff)
```

### 11.2-A Composition Comparaison using Fisher  
The odds ratio shows the degree of enrichment (>1) or depletion (<1)

```{r}
freq_fisher <- function(conditions, have_identity){
  freq <- table(factor(have_identity, levels=c(TRUE,FALSE)),
              conditions)
  test <- fisher.test(freq)
  res <- setNames(c(test$estimate, test$p.value), c("oddsratio","pval_fisher"))
  return(res)
}

region_enrichment <- data.frame(region = levels(seurat.rna$gtype),
                                t(sapply(levels(seurat.rna$gtype), function(gtype)
                                  freq_fisher(conditions = factor(seurat.rna$status, levels=c("C","D")),
                                              have_identity = seurat.rna$gtype == gtype)
                                  )),
                                row.names=NULL)
region_enrichment$padj_fisher <- p.adjust(region_enrichment$pval_fisher)
region_enrichment
```

### 11.2-B Composition Comparaison using GLM  
Fisher's exact test does not take into account replicate information. GLM is going to take account of these information

```{r}
freq_glm_aov <- function(samples, conditions, have_identity){
  sample_conditions <- unique(data.frame(sample = samples, condition = conditions))
  sample_conditions <- setNames(sample_conditions$condition, sample_conditions$sample)
  freq <- table(samples, factor(have_identity, levels=c(TRUE,FALSE)))
  m <- glm(freq ~ sample_conditions[rownames(freq)], family = "binomial")
  aov <- anova(m, test = "Chisq")
  res <- setNames(c(coef(m)[2], aov$Pr[2]), c("coef_glm","pval_aov"))
  return(res)
}

region_enrichment <- data.frame(region_enrichment,
                                t(sapply(levels(seurat.rna$gtype), function(gtype){
                                freq_glm_aov(samples = seurat.rna$orig.ident,
                                             conditions = factor(seurat.rna$status, levels=c("C","D")),
                                             seurat.rna$gtype == gtype)
                                  })),
                                row.names=NULL)
region_enrichment$padj_aov <- p.adjust(region_enrichment$pval_aov)
region_enrichment
```

Show the proportion of each region in each sample separately

```{r}
org_status <- unique(data.frame(orig.ident=seurat.rna$orig.ident,
                                status=factor(seurat.rna$status, levels=c("C","D"))))
org_status <- setNames(org_status$status, org_status$orig.ident)

layout(matrix(1:6,nrow=1))

for(gtype in c("IIx","Mix", "IIb", "D_IIb", "MTJ", "NMJ", "FAPs", "Other", "Unclassified")){
  props <- setNames(sapply(sort(unique(seurat.rna$orig.ident)), function(orig.ident)
    mean(seurat.rna$gtype[seurat.rna$orig.ident == orig.ident] == gtype)),
    sort(unique(seurat.rna$orig.ident)))
  barplot(props[order(props)],
          col = ifelse(org_status[names(props)[order(props)]] == "C", "#cdcdcd", "#303030"),
          ylab = "Proportions", main = gtype, las = 2, cex.names = 0.8)
  }

plot.new()
legend("left", fill = c("#cdcdcd","#303030"), legend = c("C","D"), bty="n")
```

##  12. Differential Expression 

### 12.1 Transcriptome similarity 
Compare the general transcriptome similarities of the same cell type/state in different samples or conditions in order to get general information whether there exists any substantial transcriptomic changes in this cell type/state between conditions.

```{r}
seurat_transcriptome <- subset(seurat.rna, subset = gtype == "IIa") %>%
  FindVariableFeatures()
avg_expr_org <- sapply(sort(unique(seurat.rna$orig.ident)), function(org)
  rowMeans(seurat_transcriptome@assays$RNA@data[,which(seurat_transcriptome$orig.ident == org)]))
corr_org <- cor(avg_expr_org[VariableFeatures(seurat_transcriptome),])
plot(hclust(as.dist(1-corr_org)))
```

### 12.2 Wilcoxon test for DE

```{r}
DE_wilcoxauc_transcriptome <- wilcoxauc(seurat_transcriptome, group_by = "status") %>%
  dplyr::filter(group == "D") %>%
  mutate(DE = abs(logFC)>log(1.1) & padj < 0.01) %>%
  mutate(DEG = ifelse(abs(logFC)>log(1.1) & padj < 0.01, feature, NA))

DE_wilcoxauc_transcriptome
write.csv(DE_wilcoxauc_transcriptome, "IPA/wilcox_IIx.csv", row.names=FALSE)

ggplot(DE_wilcoxauc_transcriptome, aes(x = logFC, y = -log10(padj), col=DE, label=DEG)) +
  geom_point() +
  geom_text_repel() +
  geom_vline(xintercept=c(-log(1.1), log(1.1), 0), col="#303030", linetype="dotted") +
  geom_hline(yintercept=-log10(0.01), col="#303030", linetype="dotted") +
  scale_color_manual(values=c("#909090", "red")) +
  theme_minimal()
```

### 12.3 DESeq2 for DE 

```{r}
det_rate <- rowMeans(seurat_transcriptome@assays$RNA@data)
meta <- seurat_transcriptome@meta.data %>% mutate(status = factor(status, levels=c("C","D")))
dds <- DESeqDataSetFromMatrix(seurat_transcriptome@assays$RNA@counts[det_rate > 0.05,],
                              colData=meta,
                              design = ~ status)

sizeFactors(dds) <- calculateSumFactors(seurat_transcriptome@assays$RNA@counts[det_rate > 0.05,])
dds <- DESeq(dds, test="LRT", reduced=~1, useT=TRUE, minmu=1e-6, minReplicatesForReplace=Inf)
DE_deseq2_transcriptome <- results(dds) %>% as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  mutate(DE = abs(log2FoldChange)>log2(1.1) & padj < 0.01) %>%
  mutate(DEG = ifelse(abs(log2FoldChange)>log2(1.1) & padj < 0.01, gene, NA))

DE_deseq2_transcriptome
write.csv(DE_deseq2_transcriptome, "IPA/DESeq_IIx.csv", row.names=FALSE)


ggplot(DE_deseq2_transcriptome, aes(x = log2FoldChange, y = -log10(padj), col=DE, label=DEG)) +
  geom_point() +
  geom_text_repel(max.overlaps = 20) +
  geom_vline(xintercept=c(-log2(1.1), log2(1.1), 0), col="#303030", linetype="dotted") +
  geom_hline(yintercept=-log10(0.01), col="#303030", linetype="dotted") +
  scale_color_manual(values=c("#909090", "red")) +
  theme_minimal()
```

### 12.3 GLM for DE 

```{r}
# Use log-normalized data as the input assumes a normal distribution. 
aov_DE <- function(expr, cond, covar = NULL, family = "gaussian", test = NULL){
  if (is.null(covar)) covar <- data.frame(X_const = 1)
  dat <- data.frame(y = expr,
                    cond = cond,
                    covar)
  m <- glm(y ~ ., family = family, data = dat)
  m0 <- glm(y ~ . - cond, family = family, data = dat)
  if (is.null(test)){
    test <- "F"
    if (ifelse(is(family, "family"), family$family, family) %in% c("binomial","poisson"))
      test <- "Chisq"
  }
  aov <- anova(m, m0, test = test)
  res <- c(coef(m)[grep("^cond",names(coef(m)))], aov$Pr[2])
  names(res)[length(res)] <- "pval"
  return(res)
}

DE_aov <- data.frame(t(sapply(which(det_rate > 0.05), function(i)
  aov_DE(expr = seurat_transcriptome@assays$RNA@data[i,],
         cond = factor(seurat_transcriptome$status, levels=c("C","D")),
         covar = data.frame(cov = log10(seurat_transcriptome$nCount_RNA))))), row.names=NULL)

DE_aov$padj <- p.adjust(DE_aov$pval, method="BH")
DE_aov <- data.frame(gene = names(which(det_rate > 0.05)), DE_aov) %>%
  mutate(DE = abs(condD)>log(1.1) & padj < 0.01) %>%
  mutate(DEG = ifelse(abs(condD)>log(1.1) & padj < 0.01, gene, NA))

DE_aov_bin <- data.frame(t(sapply(which(det_rate > 0.05), function(i)
  aov_DE(expr = seurat_transcriptome@assays$RNA@data[i,] > 0,
         cond = factor(seurat_transcriptome$status, levels=c("C","D")),
         covar = data.frame(cov = seurat_transcriptome$nCount_RNA),
         family = "binomial"))), row.names=NULL)

DE_aov_bin$padj <- p.adjust(DE_aov_bin$pval, method="BH")
DE_aov_bin <- data.frame(gene = names(which(det_rate > 0.05)), DE_aov_bin) %>%
  mutate(DE = padj < 0.01) %>%
  mutate(DEG = ifelse(padj < 0.01, gene, NA))

DE_aov
write.csv(DE_aov, "IPA/GLM_IIx.csv", row.names=FALSE)
DE_aov_bin
write.csv(DE_aov_bin, "IPA/GLMbin_IIx.csv", row.names=FALSE)


ggplot(DE_aov, aes(x = condD, y = -log10(padj), col=DE, label=DEG)) +
  geom_point() +
  geom_text_repel(max.overlaps = 20) +
  geom_vline(xintercept=c(-log(1.1), log(1.1), 0), col="#303030", linetype="dotted") +
  geom_hline(yintercept=-log10(0.01), col="#303030", linetype="dotted") +
  scale_color_manual(values=c("#909090", "red")) +
  theme_minimal()

ggplot(DE_aov_bin, aes(x = condD, y = -log10(padj), col=DE, label=DEG)) +
  geom_point() +
  geom_text_repel(max.overlaps = 20) +
  geom_vline(xintercept=0, col="#303030", linetype="dotted") +
  geom_hline(yintercept=-log10(0.01), col="#303030", linetype="dotted") +
  scale_color_manual(values=c("#909090", "red")) +
  theme_minimal()
```
### 12.4 FindMarkers for DE 

```{r}
cluster.markers <- FindMarkers(seurat.rna, ident.1 = 'C', group.by = "status", subset.ident = "IIa", min.pct = 0.25, logfc.threshold = 0.25, only.pos = F)
write.csv(cluster.markers, "IPA/FM_IIa_vs_TIIa_oP_F.csv", row.names = TRUE)
cluster.markers <- FindMarkers(seurat.rna, ident.1 = 'C', group.by = "status", subset.ident = c("IIb_1", "IIb_2", "IIb_3"), min.pct = 0.25, logfc.threshold = 0.25, only.pos = F)
write.csv(cluster.markers, "IPA/FM_IIb_vs_TIIb_oP_F.csv", row.names = TRUE)
cluster.markers <- FindMarkers(seurat.rna, ident.1 = 'C', group.by = "status", subset.ident = "IIx", min.pct = 0.25, logfc.threshold = 0.25, only.pos = F)
write.csv(cluster.markers, "IPA/FM_IIx_vs_TIIx_oP_F.csv", row.names = TRUE)

# write.csv(cluster.markers, "IPA/FM_Mix_vs_DMix_oP_T.csv", row.names = TRUE)
# cluster.markers <- FindMarkers(seurat.rna, ident.1 = 'IIb', ident.2 = "D_IIb", min.pct = 0.25, logfc.threshold = 0.25, only.pos = F)
# write.csv(cluster.markers, "IPA/FM_IIb_vs_DIIb_oP_F.csv", row.names = TRUE)
# cluster.markers <- FindMarkers(seurat.rna, ident.1 = 'IIb', ident.2 = "D_IIb", min.pct = 0.25, logfc.threshold = 0.25, only.pos = T)
# write.csv(cluster.markers, "IPA/FM_IIb_vs_DIIb_oP_T.csv", row.names = TRUE)
# cluster.markers <- FindMarkers(seurat.rna, ident.1 = c('IIx_1', 'IIx_2'), ident.2 = "D_IIx", min.pct = 0.25, logfc.threshold = 0.25, only.pos = F)
# write.csv(cluster.markers, "IPA/FM_IIx_vs_DIIx_oP_F.csv", row.names = TRUE)
# cluster.markers <- FindMarkers(seurat.rna, ident.1 = c('IIx_1', 'IIx_2'), ident.2 = "D_IIx", min.pct = 0.25, logfc.threshold = 0.25, only.pos = T)
# write.csv(cluster.markers, "IPA/FM_IIx_vs_DIIx_oP_T.csv", row.names = TRUE)
```

## 13. Trajectory Analysis 

### 13.1 Cluster connectivity analysis
PAGA approach is based on the clustering results and evaluate the strength of connectivity between every two clusters. (DOI?)

```{r}
# Convert seurat (R) to scanpy (Python) 
cell_attrs <- list(pca = Embeddings(seurat.rna,"pca")[,1:20],
                   umap = Embeddings(seurat.rna,"umap"),
                   celltype = seurat.rna@active.ident)
loom <- loomR::create("save/loom_rna.loom",
                      data = seurat.rna[['RNA']]@data,
                      layers = list(counts = seurat.rna[['RNA']]@counts),
                      cell.attrs = cell_attrs)
loom$close_all()
```

```{r}
# Convert seurat (R) to scanpy (Python) 
adata <- AnnData(X = t(seurat.rna[['RNA']]@data),
                 obs = data.frame(celltype = seurat.rna@active.ident, row.names = colnames(seurat.rna)),
                 var = seurat.rna[['RNA']]@meta.features,
                 layers = list(counts = t(seurat.rna[['RNA']]@counts)),
                 obsm = list(pca = Embeddings(seurat.rna,"pca")[,1:20],
                             umap = Embeddings(seurat.rna,"umap"))
                )
adata$write_h5ad("save/anndata_rna_final.h5ad")
```

```{r}
# Import packages
py_install("scanpy", pip=T)
py_install("igraph", pip=T)
plt <- import("matplotlib")
sc <- import("scanpy")
```

```{r}
adata.rna <- sc$read("save/anndata_rna_final.h5ad") 
# or adata_DS1 <- sc$read_loom("save/loom_rna.loom") 
sc$pp$neighbors(adata.rna, n_neighbors=20L, use_rep='pca')
sc$tl$paga(adata.rna, groups='celltype')
adata.rna$write_h5ad("save/anndata_rna_final.h5ad")
```

```{r}
plt$use("Agg", force = TRUE)
sc$pl$paga(adata.rna,
           color='celltype',
           fontsize=7,
           frameon=FALSE,
           threshold=0.2, # Change 
           save="rna_paga.png")
```

### 13.2 Using Destiny 
Pseudotime analysis

```{r}
seurat.type <- subset(seurat.rna, subset = RNA_snn_res.0.6 %in% c(0,2,3,5,6,7))
seurat.type <- FindVariableFeatures(seurat.type, nfeatures = 2000)
VariableFeatures(seurat.type) <- setdiff(VariableFeatures(seurat.type), unlist(cc.genes))
seurat.type <- CellCycleScoring(seurat.type,
                                s.features = cc.genes$s.genes,
                                g2m.features = cc.genes$g2m.genes,
                                set.ident = TRUE)
seurat.type <- ScaleData(seurat.type, vars.to.regress = c("S.Score", "G2M.Score"))
```

```{r}
seurat.type <- RunPCA(seurat.type) 
seurat.type <- RunHarmony(seurat.type, group.by.vars = "orig.ident", dims.use = 1:25, max.iter.harmony = 50)
seurat.type <- RunUMAP(seurat.type, reduction = "harmony", dims = 1:25)
seurat.type <- FindNeighbors(seurat.type, dims = 1:ncol(Embeddings(seurat.rna)), reduction = "harmony")
seurat.type <- FindClusters(seurat.type, resolution = 0.8, algorithm = 2)
DimPlot(seurat.type, reduction = "umap", group.by = "orig.ident", cols = matisse) & NoAxes()
DimPlot(seurat.type, reduction = "umap", group.by = "status", cols = c("#FFA07A", "#3F51B5")) & NoAxes()
DimPlot(seurat.type, reduction = "umap") & NoAxes()

DefaultAssay(seurat.type) <- "RNA"
input_data.markers <- FindAllMarkers(seurat.type, min.pct = 0.25, logfc.threshold = 0.25, only.pos = TRUE)
input_data.markers %>% group_by(cluster) %>% slice_max(n = 20, order_by = avg_log2FC)
input_data.markers %>% group_by(cluster) %>% top_n(n = 20, wt = avg_log2FC) -> topX
input_data.markers
FeaturePlot(seurat.type, c("Myh2", "Myh4", "Runx1", "Dlg2"), ncol = 2)
```
```{r}
FeaturePlot(seurat.type, c("Myh2", "Myh4", "Runx1", "Uty"), ncol = 2) & NoAxes()
FeaturePlot(seurat.type, c("Myh1", "Myh7", "Myh9", "Gm26917"), ncol = 2) & NoAxes()
FeaturePlot(seurat.type, c("Mast4", "Cd36", "Actn2", "Xist"), ncol = 2) & NoAxes()
```

```{r}
dm <- DiffusionMap(Embeddings(seurat.type, "pca")[,1:20])
dpt <- DPT(dm)
seurat.type$dpt <- rank(dpt$dpt) # or (-dpt$dpt)
FeaturePlot(seurat.type, c("dpt", "Myh4", "Runx1", "Dlg2"), ncol=2)
```

Visualize expression changes along the constructed pseudotime

```{r}
plot1 <- qplot(seurat.type$dpt, as.numeric(seurat.type@assays$RNA@data["Myh4",]),
               xlab="Dpt", ylab="Expression", main="Myh4") +
  geom_smooth(se = FALSE, method = "loess") + theme_bw()
plot2 <- qplot(seurat.type$dpt, as.numeric(seurat.type@assays$RNA@data["Runx1",]),
               xlab="Dpt", ylab="Expression", main="Runx1") +
  geom_smooth(se = FALSE, method = "loess") + theme_bw()
plot3 <- qplot(seurat.type$dpt, as.numeric(seurat.type@assays$RNA@data["Cntfr",]),
               xlab="Dpt", ylab="Expression", main="Cntfr") +
  geom_smooth(se = FALSE, method = "loess") + theme_bw()
plot1
plot2
plot3
```

### 13.3 Using Monocle3

```{r}
seurat.rna.cds <- as.cell_data_set(seurat.rna, reduction = 'umap')
seurat.rna.cds <- cluster_cells(cds = seurat.rna.cds, reduction_method = "UMAP")
seurat.rna.cds <- learn_graph(seurat.rna.cds, use_partition = TRUE)
seurat.rna.cds <- order_cells(seurat.rna.cds, reduction_method = "UMAP", root_cells = NULL)

plot_cells(seurat.rna.cds, color_cells_by = "status", cell_size = 0.5, group_label_size = 3, label_groups_by_cluster = F, label_leaves = FALSE, label_branch_points = FALSE, graph_label_size = 5)
plot_cells(seurat.rna.cds, color_cells_by = "gtype", cell_size = 0.5, label_groups_by_cluster = FALSE, label_leaves = FALSE, label_branch_points = FALSE, graph_label_size = 4)

rowData(seurat.rna.cds)$gene_name <- rownames(seurat.rna.cds)
rowData(seurat.rna.cds)$gene_short_name <- rowData(seurat.rna.cds)$gene_name

plot_cells(seurat.rna.cds, color_cells_by = "pseudotime", cell_size = 0.5, label_groups_by_cluster = FALSE, label_leaves = FALSE, label_branch_points = FALSE, graph_label_size = 4)
plot_cells(seurat.rna.cds, 
           genes = c("Myh4", "Runx1"),
           color_cells_by = "pseudotime", cell_size = 0.5, label_groups_by_cluster = FALSE, label_leaves = FALSE, label_branch_points = FALSE, graph_label_size = 4)
```

```{r}
subset.cds <- choose_cells(seurat.rna.cds)

plot_cells(subset.cds, genes=c("Myh1", "Myh2", "Myh4", "Runx1"),
           show_trajectory_graph=FALSE,
           label_cell_groups=FALSE,
           label_leaves=FALSE)

subset.cds.genes <- subset.cds[rowData(subset.cds)$gene_short_name %in% c("Myh1", "Myh2", "Myh4", "Runx1")]

plot_genes_in_pseudotime(subset.cds.genes,
                         color_cells_by="gtype",
                         min_expr=0.5)
```

## 14. NicheNet analysis 

_WIP_

```{r}
seurat.rna <- readRDS("save/GSE183802_snRNA-seq_data_CTL_and_denervation.rds")
seurat.rna$celltype <- Idents(seurat.rna)
# TypeI TypeIIA TypeIIx TypeIIB MTJ NMJ Satellite FAPs Pericyte Adipocyte Endothelial Macrophage Unclassified
```

```{r}
ligand_target_matrix = readRDS(url("https://zenodo.org/record/3260758/files/ligand_target_matrix.rds"))
lr_network = readRDS(url("https://zenodo.org/record/3260758/files/lr_network.rds"))
weighted_networks = readRDS(url("https://zenodo.org/record/3260758/files/weighted_networks.rds"))
weighted_networks_lr = weighted_networks$lr_sig %>% inner_join(lr_network %>% distinct(from,to), by = c("from","to"))

lr_network = lr_network %>% mutate(from = convert_human_to_mouse_symbols(from), to = convert_human_to_mouse_symbols(to)) %>% drop_na()
colnames(ligand_target_matrix) = ligand_target_matrix %>% colnames() %>% convert_human_to_mouse_symbols()
rownames(ligand_target_matrix) = ligand_target_matrix %>% rownames() %>% convert_human_to_mouse_symbols()

ligand_target_matrix = ligand_target_matrix %>% .[!is.na(rownames(ligand_target_matrix)), !is.na(colnames(ligand_target_matrix))]

weighted_networks_lr = weighted_networks_lr %>% mutate(from = convert_human_to_mouse_symbols(from), to = convert_human_to_mouse_symbols(to)) %>% drop_na()
```

```{r}
receiver = "TypeIIA"
expressed_genes_receiver = get_expressed_genes(receiver, seurat.rna, pct = 0.10, assay_oi = "RNA")
background_expressed_genes = expressed_genes_receiver %>% .[. %in% rownames(ligand_target_matrix)]
sender_celltypes = "TypeIIB"
list_expressed_genes_sender = sender_celltypes %>% unique() %>% lapply(get_expressed_genes, seurat.rna, 0.10, assay_oi = "RNA") # lapply to get the expressed genes of every sender cell type separately here
expressed_genes_sender = list_expressed_genes_sender %>% unlist() %>% unique()
```

```{r}
seurat_obj_receiver= subset(seurat.rna, idents = receiver)
seurat_obj_receiver = SetIdent(seurat_obj_receiver, value = seurat_obj_receiver[["sti"]])
condition_oi = "N"
condition_reference = "C" 
DE_table_receiver = FindMarkers(object = seurat_obj_receiver, ident.1 = condition_oi, ident.2 = condition_reference, min.pct = 0.10) %>% rownames_to_column("gene")
geneset_oi = DE_table_receiver %>% filter(p_val_adj <= 0.05 & abs(avg_log2FC) >= 0.25) %>% pull(gene)
geneset_oi = geneset_oi %>% .[. %in% rownames(ligand_target_matrix)]
```

```{r}
ligands = lr_network %>% pull(from) %>% unique()
receptors = lr_network %>% pull(to) %>% unique()
expressed_ligands = intersect(ligands,expressed_genes_sender)
expressed_receptors = intersect(receptors,expressed_genes_receiver)
potential_ligands = lr_network %>% filter(from %in% expressed_ligands & to %in% expressed_receptors) %>% pull(from) %>% unique()
```

```{r}
ligand_activities = predict_ligand_activities(geneset = geneset_oi, background_expressed_genes = background_expressed_genes, ligand_target_matrix = ligand_target_matrix, potential_ligands = potential_ligands)
ligand_activities = ligand_activities %>% arrange(-pearson) %>% mutate(rank = rank(desc(pearson)))
ligand_activities
```

```{r}
best_upstream_ligands = ligand_activities %>% top_n(20, pearson) %>% arrange(-pearson) %>% pull(test_ligand) %>% unique()
active_ligand_target_links_df = best_upstream_ligands %>% lapply(get_weighted_ligand_target_links,geneset = geneset_oi, ligand_target_matrix = ligand_target_matrix, n = 200) %>% bind_rows() %>% drop_na()
active_ligand_target_links = prepare_ligand_target_visualization(ligand_target_df = active_ligand_target_links_df, ligand_target_matrix = ligand_target_matrix, cutoff = 0.33)
order_ligands = intersect(best_upstream_ligands, colnames(active_ligand_target_links)) %>% rev() %>% make.names()
order_targets = active_ligand_target_links_df$target %>% unique() %>% intersect(rownames(active_ligand_target_links)) %>% make.names()
rownames(active_ligand_target_links) = rownames(active_ligand_target_links) %>% make.names() # make.names() for heatmap visualization of genes like H2-T23
colnames(active_ligand_target_links) = colnames(active_ligand_target_links) %>% make.names() # make.names() for heatmap visualization of genes like H2-T23
vis_ligand_target = active_ligand_target_links[order_targets,order_ligands] %>% t()
p_ligand_target_network = vis_ligand_target %>% make_heatmap_ggplot("Prioritized ligands","Predicted target genes", color = "purple",legend_position = "top", x_axis_position = "top",legend_title = "Regulatory potential")  + theme(axis.text.x = element_text(face = "italic")) + scale_fill_gradient2(low = "whitesmoke",  high = "purple", breaks = c(0,0.0045,0.0090))
```


